cmd/core/pdr.go

```diff
@@ -0,0 +1,136 @@
+package core
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/edgecomllc/eupf/cmd/ebpf"
+	"github.com/rs/zerolog/log"
+	"github.com/wmnsk/go-pfcp/ie"
+)
+
+func deletePDR(spdrInfo SPDRInfo, mapOperations ebpf.ForwardingPlaneController, pdrCtx *PDRCreationContext) error {
+	if spdrInfo.Ipv4 != nil {
+		if err := mapOperations.DeletePdrDownlink(spdrInfo.Ipv4); err != nil {
+			return fmt.Errorf("Can't delete IPv4 PDR: %s", err.Error())
+		}
+	} else if spdrInfo.Ipv6 != nil {
+		if err := mapOperations.DeleteDownlinkPdrIp6(spdrInfo.Ipv6); err != nil {
+			return fmt.Errorf("Can't delete IPv6 PDR: %s", err.Error())
+		}
+	} else {
+		if err := mapOperations.DeletePdrUplink(spdrInfo.Teid); err != nil {
+			return fmt.Errorf("Can't delete GTP PDR: %s", err.Error())
+		}
+	}
+	if spdrInfo.Teid != 0 {
+		pdrCtx.ResourceManager.FTEIDM.ReleaseTEID(pdrCtx.Session.RemoteSEID)
+	}
+	return nil
+}
+
+func extractPDR(pdr *ie.IE, spdrInfo *SPDRInfo, pdrContext *PDRCreationContext) error {
+	if outerHeaderRemoval, err := pdr.OuterHeaderRemovalDescription(); err == nil {
+		spdrInfo.PdrInfo.OuterHeaderRemoval = outerHeaderRemoval
+	}
+	if farid, err := pdr.FARID(); err == nil {
+		spdrInfo.PdrInfo.FarId = pdrContext.getFARID(farid)
+	}
+	if qerid, err := pdr.QERID(); err == nil {
+		spdrInfo.PdrInfo.QerId = pdrContext.getQERID(qerid)
+	}
+
+	pdi, err := pdr.PDI()
+	if err != nil {
+		return fmt.Errorf("PDI IE is missing")
+	}
+
+	if sdfFilter, err := pdr.SDFFilter(); err == nil {
+		if sdfFilterParsed, err := ParseSdfFilter(sdfFilter.FlowDescription); err == nil {
+			spdrInfo.PdrInfo.SdfFilter = &sdfFilterParsed
+		} else {
+			log.Error().Msgf("SDFFilter err: %v", err)
+			return err
+		}
+	}
+
+	//Bug in go-pfcp:
+	//if fteid, err := pdr.FTEID(); err == nil {
+	if teidPdiId := findIEindex(pdi, 21); teidPdiId != -1 { // IE Type F-TEID
+		if fteid, err := pdi[teidPdiId].FTEID(); err == nil {
+			var teid = fteid.TEID
+			if pdrContext.ResourceManager != nil && pdrContext.ResourceManager.FTEIDM != nil {
```

Comment:
"Train" calls... Please provide single direct method to check whether we can allocate smthig or not.

Like `pdrContext.canDoIt()`